### ISSUE TITLE
step crypto nacl box only supports string nonces

### ISSUE THREAD

**Issue by jumaffre** [smallstep/cli#279]
### What would you like to be added

`step crypto nacl box` CLI currently only supports string nonces. As opposed to the input/output that can be base64-encoded or raw (using the `-raw` option), the nonce _has_ to be a string. 

However, in some cases, a secret could have been sealed by a third-party system that generated the nonce at random. This nonce could contain null bytes. In this case, passing the nonce to `step` during `box open` is not straightforward.

Instead, it would be awesome if there was an option to specify that the input nonce is base64-encoded (probably different than the existing `-raw` option).

### Why this is needed

Currently working on https://github.com/Microsoft/CCF, we are trying to provide users with examples of third-party applications they can use to decrypt their recovery shares. `step` seems like a very straightforward CLI that we would love to refer to and use in our sample scripts.

**Comment by maraino** [comment]
@jumaffre: it's a good idea, and I'm gonna implement it for all nacl commands that require a nonce, what I'm not sure yet if I should add a flag or do something like:
```
cat data.enc | step crypto nacl box open base64:MTIz sender.pub box.priv
```

I'm not sure how to name a flag that indicates the nonce is in base64 at least something better than just `base64:<base64-encoded nonce>`. Do you have a better suggestion?

**Comment by jumaffre** [comment]
@maraino Awesome! Thanks for implementing this feature so quickly. The current `base64:` prefix seems sound to me, except perhaps that `base64:` cannot be used as a string nonce itself. I am not familiar with the philosophy of the step CLI, but what about adding an extra `-b64-nonce` argument?

Do you roughly know when the next release with this feature in will be rolled out?

**Comment by maraino** [comment]
Yes, I thought about that, `base64:` can only be encoded as `base64:YmFzZTY0Og==`. I thought it was a good trade-off.

I can add the extra `--b64-nonce`. Regarding the next version, there has been some changes in master, so I guess we'll release a minor version during this week.

**Comment by mmalone** [comment]
Sorry I'm coming to this late.

This is a sort of funky situation and I can't think of any precedence in the CLI that would dictate style here. I don't hate the `base64:XXX` solution that @mariano proposed. To address the issue about encoding `base64:` we could also add the `string:` prefix and make it the default. I imagine the logic would be:

* If the nonce doesn't have a `:` in it then the format defaults to `string`
* If the nonce has a `:` in it the format is indicated by the string to the left of the first `:` (so `XXX:YYY:ZZZ:QQQ` would have format `XXX` and nonce `YYY:ZZZ:QQQ`)

So you could encode `base64:` as `string:base64:`. If you want to be explicit, you'd _always_ want to specify your nonce-type prefix, but for the normal use case where the nonce is probably a random alphanumeric string the default `string:` type would work. The nice thing about defining this way is that it leaves room for other formats in the future.

I'm also ok with using a flag for this. But, to address that last point about leaving room for future formats, it feels like this should be something like `--nonce-format <format>` or `--nonce-type <type>` instead of `--b64-nonce`. Again, the default format/type would be string. This way we don't have to add more flags if we add new formats/types and we don't have to deal with special error checking for mutually exclusive flags like `--string-nonce --b64-nonce`.

I'm sort of ambivalent regarding which direction we go with here -- unless someone has a good argument I think it's implementor's choice -- but I think we should choose one or the other: either the prefix or the flag. Having two ways to do the same thing seems unnecessary and potentially confusing, and we'd need to handle the error cases where someone writes `step crypto nacl box open string:foo x.pub x.priv --b64-format`.

**Comment by maraino** [comment]
@jumaffre @mmalone: I've added the `string:` prefix too, and at least for the moment we're not gonna add the flags.

Although we're not using prefixes yet in `step`,  I am planning to add them when we add support for HSMs in the cli. For example, it might be possible to use your youbikey using things like `yubikey:slotId` adding support for flags for multiple HSM implementations will force us to add a large number of flags, especially when they can be encoded in a uri-like format.

**Comment by jumaffre** [comment]
@maraino Just a ping to ask whether the next release of the step CLI is planned soon? We are super keen to provide a wrapper for it for CCF users. Thanks!

**Comment by maraino** [comment]
@jumaffre I'll make a task for releasing on Monday. I'm sorry we didn't do it before.

**Comment by maraino** [comment]
Hi @jumaffre just letting you know, that I've just released v0.14.4
https://github.com/smallstep/cli/releases/tag/v0.14.4

And I'm in the process of releasing it to other distribution channels.

**Comment by jumaffre** [comment]
@maraino Awesome! Thanks very much for implementing this so quickly. I will test this out by the end of this week 👍
